src/main/host/descriptor/epoll.c

```diff
@@ -523,11 +566,30 @@ gint epoll_getEvents(Epoll* epoll, struct epoll_event* eventArray, gint eventArr
      * overflow. the number of actual events is returned in nEvents. */
     gint eventIndex = 0;
 
-    GHashTableIter iter;
-    gpointer key, value;
-    g_hash_table_iter_init(&iter, epoll->ready);
-    while(g_hash_table_iter_next(&iter, &key, &value) && (eventIndex < eventArrayLength)) {
-        EpollWatch* watch = value;
+    /*
+     * We need to guarantee that the events are returned in a determinstic order when the
+     * simulation is run multiple times, so we cannot use hash table iterator.
+     * Using a list here has some potential performance implications:
+     * - O(n) to loop the hash table and create the list of values
+     * - O(n) to sort the list
+     * - O(n) for our iteration of the list
+     * We think that the ready list is typically small and so 3*O(n) will be small in practice.
+     * If this turns out not to be the case, we could consider maintaining a sorted list of the
+     * ready watch values alongside the epoll->ready hash table instead, which would allow us
+     * to reduce this function to 1*O(n) by avoiding creating and sorting the list here.
```

Comment:
Again might be worth mentioning or using https://developer.gnome.org/glib/2.68/glib-Balanced-Binary-Trees.html

Reply:
Added a note to consider trees. Since this PR appears to have a negligible performance impact, let's leave enhancements to epoll etc. for a future PR (maybe when we rewrite in rust).